FontSizePicker: Refactor stories to use Controls

mirka opened the pull request:
Part of #35665
In preparation for #38720 

## Description

Refactors the `FontSizePicker` stories to use Controls instead of Knobs, and reworks some of the stories for clarity and correctness.

### Reworked stories

- "With Slider"
  - [Bug fix] The `fallbackFontSize` prop only kicks in when `initialValue` is `0` or `undefined`, but this was not set correctly so it wasn't showcasing the fallback behavior.
- "Different Control By Size"
  - This was split into two stories ("With More Font Sizes" and "With Units") for clarity.
  - [Bug fix] The "With Units" case was setting an `initialValue` of `8` instead of `8px`, thus showcasing a custom value by default.
- "With Complex CSS Values"
  - Simplified so it doesn't require Storybook-only toggles.

I tried not to drop any of the intent captured in the original stories — let me know if I missed something.

## Testing Instructions

1. `npm run storybook:dev` and see the stories for `FontSizePicker`.
2. Compare with the ones on https://wordpress.github.io/gutenberg/?path=/story/components-fontsizepicker--default

Pro tip: Knobs can be kind of buggy when you click through multiple stories. Click the "Reset" button in the bottom right corner if they get stuck.

## Types of changes

Storybook only.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [x] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

mirka committed:
FontSizePicker: Refactor stories to use Controls

ciampo reviewed (commented):
Thank you @mirka , this is a great refactor!

Not directly a consequence of this PR, but I got confused when toggling the `withReset` prop — even when `true`, [the reset button is only shown when enabling a custom size while not showing a slider](https://github.com/WordPress/gutenberg/blob/aac3e86b3858143c967bfcc28ab11f72d3aafa1b/packages/components/src/font-size-picker/index.js):

- the docs for the prop are not very clear to me: "_If `true`, a reset button will be displayed alongside the predefined and custom font size fields._". Should we update them?
- We should probably add a similar sentence as a note to the Storybook toggle, to avoid further confusion

Also looking forward to having @ntsekouras and @aaronrobertshaw take a look at this too.

ntsekouras reviewed (commented):
Thanks for the PR @mirka !

>the docs for the prop are not very clear to me: "If true, a reset button will be displayed alongside the predefined and custom font size fields.". Should we update them?

Yes, we should do that! It through me off at first and had to carefully read the code again to see what is expected and what's not 😄 

In general I like the direction of this, but I'm wondering if we should revisit what stories we should show now. Previously with the knobs, not all properties were editable and were there to show case some combinations - thus @ciampo 's confusion (and mine) with the `reset` or `disableCustomFontSizes`, etc..

I think we can reduce the stories shown here and combine them by changing some names. For example `WithoutCustomSizes` story is identical with the `default` except on property, but by being able to edit the props it starts being confusing. 

Maybe it makes sense to always show the `fewer/more` font sizes? 🤔 
Alternatively we could [disable some controls](https://storybook.js.org/docs/react/essentials/controls#disable-controls-for-specific-properties) to mimic the previous behavior. 

What do you think?

aaronrobertshaw reviewed (commented):
Thanks for getting this refactor started @mirka 👍 

All the font size picker examples work as advertised for me.

@ntsekouras raises some good points though. I think for a lot of components we could combine examples however for others it makes sense to demonstrate a certain configuration and behaviour. Allowing all props to be edited in these scenarios definitely adds confusion and doesn't seem right.

I'd also like to see some examples being clearer in what they are supposed to demonstrate. One example of this could be "With Complex CSS Values". I'm not sure someone looking at these for the first time would pick that the difference between this example and "With Units" is the labelling of the segmented control. Could we communicate this better within the example even if just with some help text below it explaining/highlighting the demonstrated behaviour?

These aren't issues specifically with this PR of course however it would be good to form a consensus plan here so it can be followed for other components.

mirka committed:
Fix docs for `withReset`

mirka committed:
Expand code snippets and show descriptions in Controls

mirka committed:
Tweak `withReset` setup for better code snippets

These can be removed once the component is TS converted

mirka committed:
Add manual descriptions for confusing props

These can be removed once the component is TS converted

mirka committed:
Add descriptions to confusing stories

This would be more convenient https://github.com/izhan/storybook-description-loader

mirka commented:
Thanks for all the feedback, you all raise very good points — the main sentiment being that it's hard to understand what the props do or what the stories are trying to demonstrate. I went back and did some thinking on how we should be approaching Storybook, and the ideas are discussed in #38842.

Based on that, I think the direction we’ll be pursuing is to make Docs view (instead of Canvas view) the primary focus of   our Storybook. Once a component is written in TypeScript, the docgen will not only extract all the prop data into a table, but also **auto-generate the Controls for every prop**. That is super convenient, but also that alone would perhaps make some stories overwhelming to understand.

However, in Docs view we can show descriptions and actual code snippets for each story, which is actually more useful to most devs than having just a curated set of Controls to tinker with. The Canvas view would instead be more of a place to do advanced tinkering, with all the Controls present.

While `FontSizePicker` is not in TypeScript yet, I made some changes with that eventuality in mind:

- Make code snippets visible in Docs view, and checked that they were understandable.
- Manually added descriptions to confusing props. (This will be automatic when the component is converted to TS)
- Added descriptions to some stories to clarify what they were trying to demonstrate.

How does it feel now?

https://user-images.githubusercontent.com/555336/154745860-9e706257-005f-4c99-967b-419325e62160.mp4

mirka commented on packages/components/src/font-size-picker/README.md:
```diff
@@ -100,8 +100,7 @@ If `true`, the UI will contain a slider, instead of a numeric text input field.
 
-If `true`, a reset button will be displayed alongside the predefined and custom
-font size fields.
+If `true`, a reset button will be displayed alongside the input field when a custom font size is active. Has no effect when `disableCustomFontSizes` or `withSlider` is `true`.
```
Fixed, thanks @ciampo for noticing.

mirka commented on packages/components/src/font-size-picker/stories/index.js:
```diff
@@ -71,45 +78,37 @@ export const withSlider = () => {
+				'When disabled, the user will only be able to pick one of the predefined sizes passed in `fontSizes`.',
 		},
-	] );
-	return (
-		<FontSizePickerWithState
-			fontSizes={ fontSizes }
-			disableCustomFontSizes
-		/>
-	);
+	},
```
This is kind of verbose for just wanting to set a story description, so I'm thinking of using a [webpack loader](https://github.com/izhan/storybook-description-loader) that lets you write the description as a normal JSDoc comment.

ciampo replied:
Do you think there's any chance of breaking our story description in case Storybook updates their APIs? The webpack loader that you linked to hasn't been updated in more than a year

Anyway, we can definitely experiment with it in a separate PR! (although definitely not high priority as it's just a minor improvement)

mirka replied:
The loader code is pretty trivial and there's a [reasonable chance](https://github.com/storybookjs/storybook/issues/8527) that it will become a built-in feature, so I'm not _too_ concerned about that. I might even try keeping the loader code local, rather than as an external dep.

ciampo commented on packages/components/src/font-size-picker/stories/index.js:
```diff
@@ -71,45 +78,37 @@ export const withSlider = () => {
+	docs: {
+		description: {
+			story:
+				'When disabled, the user will only be able to pick one of the predefined sizes passed in `fontSizes`.',
```
```suggestion
				'With custom font sizes disabled via the `disableCustomFontSizes` prop, the user will only be able to pick one of the predefined sizes passed in `fontSizes`.',
```

ciampo reviewed (approved):
> I think the direction we’ll be pursuing is to make Docs view (instead of Canvas view) the primary focus of our Storybook.
> ...
> The Canvas view would instead be more of a place to do advanced tinkering, with all the Controls present.

I also think that this is the right direction to take

> Once a component is written in TypeScript, the docgen will not only extract all the prop data into a table, but also auto-generate the Controls for every prop. That is super convenient, but also that alone would perhaps make some stories overwhelming to understand.

Those considerations are all correct, and (as discussed in https://github.com/WordPress/gutenberg/pull/38842) I think we can mitigate the potentially overwhelming amount of props / controls in a few ways (e.g. sorting, grouping..)

> While FontSizePicker is not in TypeScript yet, I made some changes with that eventuality in mind:
> - Make code snippets visible in Docs view, and checked that they were understandable.
> - Manually added descriptions to confusing props. (This will be automatic when the component is converted to TS)
> - Added descriptions to some stories to clarify what they were trying to demonstrate.
> 
> How does it feel now?

I think it's much better already! Although it's probably wiser to hear at least @ntsekouras 's opinion too before merging (I know Aaron may be AFK for a while)

Also, is it expected that the header hint in the "With Units" Story only has the unit (`px`) in the toggle-group version of the picker, but has the quantity + the unit (`12px`) in the dropdown-version?

![Screenshot 2022-02-23 at 19 10 54](https://user-images.githubusercontent.com/1083581/155380931-335aa1c8-56b7-4f2f-bbea-94ee861f5eeb.png)

Finally — should we mention the stories rework (and the couple of bug fixes) in a CHANGELOG entry?

ntsekouras commented:
>Also, is it expected that the header hint in the "With Units" Story only has the unit (px) in the toggle-group version of the picker, but has the quantity + the unit (12px) in the dropdown-version?

Yes. The same information is displayed in different places (hint, labels, etc..).

ntsekouras reviewed (approved):
Thanks @mirka! That's so much better now with all the descriptions and hints! Great work 🚀

mirka committed:
Improve story description

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>
